frontend/src/components/viewers/MetricsVisualizations.tsx

```diff
@@ -344,14 +362,227 @@ type ConfidenceMetric = {
-  const confidenceMetricsDataList = artifacts
-    .map(artifact => {
+// Get the columns, rows, and id-color mappings for filter functionality.
+const getRocCurveFilterTable = (
+  confidenceMetricsDataList: ConfidenceMetricsData[],
+  linkedArtifactsPage: LinkedArtifact[],
+  filter?: ConfidenceMetricsFilter,
+): RocCurveFilterTable => {
+  const columns: Column[] = [
+    {
+      customRenderer: executionArtifactCustomRenderer,
+      flex: 1,
+      label: 'Execution name > Artifact name',
+    },
+    {
+      customRenderer: runNameCustomRenderer,
+      flex: 1,
+      label: 'Run name',
+    },
+    { customRenderer: curveLegendCustomRenderer, flex: 1, label: 'Curve legend' },
+  ];
+  const rows: TableRow[] = [];
+  if (filter) {
+    const { selectedIds, selectedIdColorMap, setSelectedIdColorMap, fullArtifactPathMap } = filter;
+
+    // Only display the selected ROC Curves on the plot, in order of selection.
+    const confidenceMetricsDataMap = new Map();
+    for (const confidenceMetrics of confidenceMetricsDataList) {
+      confidenceMetricsDataMap.set(confidenceMetrics.id, confidenceMetrics);
+    }
+    confidenceMetricsDataList = selectedIds.map(selectedId =>
+      confidenceMetricsDataMap.get(selectedId),
+    );
+
+    // Populate the color map on the initial render.
+    if (selectedIds.length > 0 && Object.keys(selectedIdColorMap).length === 0) {
+      selectedIds.forEach(selectedId => {
+        selectedIdColorMap[selectedId] = lineColorsStack.pop()!;
+      });
+      setSelectedIdColorMap(selectedIdColorMap);
+    }
+
+    // Populate the filter table rows.
+    for (const linkedArtifact of linkedArtifactsPage) {
+      const id = getRocCurveId(linkedArtifact);
+      const fullArtifactPath = fullArtifactPathMap[id];
+      const row = {
+        id,
+        otherFields: [
+          `${fullArtifactPath.execution.name} > ${fullArtifactPath.artifact.name}`,
+          fullArtifactPath.run,
+          selectedIdColorMap[id],
+        ] as any,
+      };
+      rows.push(row);
+    }
+  }
+  return {
+    columns,
+    rows,
+    selectedConfidenceMetrics: confidenceMetricsDataList,
+  };
+};
+
+const updateRocCurveSelection = (
+  filter: ConfidenceMetricsFilter,
+  maxSelectedRocCurves: number,
+  newSelectedIds: string[],
+): void => {
+  const {
+    selectedIds: oldSelectedIds,
+    setSelectedIds,
+    selectedIdColorMap,
+    setSelectedIdColorMap,
+  } = filter;
+
+  // Convert arrays to sets for quick lookup.
+  const newSelectedIdsSet = new Set(newSelectedIds);
+  const oldSelectedIdsSet = new Set(oldSelectedIds);
+
+  // Find the symmetric difference and intersection of new and old IDs.
+  const addedIds = newSelectedIds.filter(selectedId => !oldSelectedIdsSet.has(selectedId));
+  const removedIds = oldSelectedIds.filter(selectedId => !newSelectedIdsSet.has(selectedId));
+  const sharedIds = oldSelectedIds.filter(selectedId => newSelectedIdsSet.has(selectedId));
```

Comment:
This filtering logic runs in `O(n)` of the number of artifacts, as the `newSelectedIds` count could be all of the artifacts present. I use the `Set` conversion to make the `filter` process faster (rather than a potential `O(n^2)` runtime).

These `addedIds`, `removedIds`, and `sharedIds` variables are used below to carefully restrict which new artifacts are actually selected (as we only allow a maximum of `10` selected ROC Curves), and assign those selected artifacts to colors from the `lineColorsStack`.